src/coreclr/jit/ssabuilder.cpp

```diff
@@ -1198,6 +1198,7 @@ void SsaBuilder::RenameVariables()
         assert(varDsc->lvTracked);
 
         if (varDsc->lvIsParam || m_pCompiler->info.compInitMem || varDsc->lvMustInit ||
+            (varTypeIsGC(varDsc) && !varDsc->lvHasExplicitInit) ||
             VarSetOps::IsMember(m_pCompiler, m_pCompiler->fgFirstBB->bbLiveIn, varDsc->lvVarIndex))
```

Comment:
I was hitting an assertion in SSA where `optRemoveRedundantZeroInits` was removing a zero init based on the knowledge that it will be inited in the prolog, while the logic in SSA and VN disagreed about this fact (and thus didn't add any initial SSA value). The problem seems to be that this condition of `fgVarNeedsExplicitZeroInit` is not reflected by SSA or VN in any way:

https://github.com/dotnet/runtime/blob/21f356a4f955bcff66f84762be4826ba605a3645/src/coreclr/jit/compiler.hpp#L4201-L4226

The zero init that `optRemoveRedundantZeroInits` removes looks like:
```scala
*************** In optRemoveRedundantZeroInits()

removing useless STMT00018 ( INL04 @ 0x000[E-] ... ??? ) <- INL03 @ ??? <- INLRT @ 0x013[E-]
N002 (  5,  6) [000067] DA---------                         ▌  STORE_LCL_FLD ref   (P) V03 tmp3         [+0]
                                                            ▌    ref    field V03._source (fldOffset=0x0) -> V12 tmp12        
N001 (  1,  1) [000057] -----------                         └──▌  CNS_INT   ref    null
 from BB01
```
However, SSA/VN do not add an initial value for `V12` and we hit an SSA assert on a subsequent use of `V12`.

The fix here uses the knowledge that all `TYP_REF` locals without explicit inits are guaranteed to be initialized in the prolog, so they have an initial value.

Reply:
BTW, there is also a work item here to ensure that local morph folds struct fields in post order as well to avoid DNER in a case like the above. The store ought to be folded to be to `V12` directly. I will look into that as a follow-up.